api-get.go

```diff
@@ -130,122 +131,193 @@ func (c Client) GetBucketACL(bucketName string) (BucketACL, error) {
-	offset     int64
-	bucketName string
-	objectName string
+	// User allocated and defined.
+	reqCh      chan<- readAtRequest
+	resCh      <-chan readAtResponse
+	doneCh     chan<- struct{}
+	objectSize int64
+
+	// Previous error saved for future calls.
+	prevErr error
 }
 
-// newObjectReadSeeker wraps getObject request returning a io.ReadSeeker.
-func newObjectReadSeeker(client Client, bucket, object string) *objectReadSeeker {
-	return &objectReadSeeker{
+// newObjectReadAtCloser implements a io.ReadSeeker for a HTTP stream.
+func newObjectReadAtCloser(reqCh chan<- readAtRequest, resCh <-chan readAtResponse, doneCh chan<- struct{}, objectSize int64) *objectReadAtCloser {
+	return &objectReadAtCloser{
 		mutex:      new(sync.Mutex),
-		reader:     nil,
-		isRead:     false,
-		client:     client,
-		offset:     0,
-		bucketName: bucket,
-		objectName: object,
+		reqCh:      reqCh,
+		resCh:      resCh,
+		doneCh:     doneCh,
+		objectSize: objectSize,
 	}
 }
 
-// Read reads up to len(p) bytes into p.  It returns the number of bytes
-// read (0 <= n <= len(p)) and any error encountered.  Even if Read
-// returns n < len(p), it may use all of p as scratch space during the call.
-// If some data is available but not len(p) bytes, Read conventionally
-// returns what is available instead of waiting for more.
+// ReaderAt is the interface that wraps the basic ReadAt method.
+//
+// ReadAt reads len(p) bytes into p starting at offset off in the
+// underlying input source.  It returns the number of bytes
+// read (0 <= n <= len(p)) and any error encountered.
 //
-// When Read encounters an error or end-of-file condition after
-// successfully reading n > 0 bytes, it returns the number of
-// bytes read.  It may return the (non-nil) error from the same call
-// or return the error (and n == 0) from a subsequent call.
-// An instance of this general case is that a Reader returning
-// a non-zero number of bytes at the end of the input stream may
-// return either err == EOF or err == nil.  The next Read should
-// return 0, EOF.
-func (r *objectReadSeeker) Read(p []byte) (int, error) {
+// When ReadAt returns n < len(p), it returns a non-nil error
+// explaining why more bytes were not returned.  In this respect,
+// ReadAt is stricter than Read.
+//
+// Even if ReadAt returns n < len(p), it may use all of p as scratch
+// space during the call.  If some data is available but not len(p) bytes,
+// ReadAt blocks until either all the data is available or an error occurs.
+// In this respect ReadAt is different from Read.
+//
+// If the n = len(p) bytes returned by ReadAt are at the end of the
+// input source, ReadAt may return either err == EOF or err == nil.
+//
+// If ReadAt is reading from an input source with a seek offset,
+// ReadAt should not affect nor be affected by the underlying
+// seek offset.
+//
+// Clients of ReadAt can execute parallel ReadAt calls on the
+// same input source.
+//
+// Implementations must not retain p.
```

Comment:
This is official defn. Let ReadAt always set error = io.EOF at EOF, even when n = len(b).

ReadAt reads len(b) bytes from the File starting at byte offset off. It returns the number of bytes read and the error, if any. ReadAt always returns a non-nil error when n < len(b). At end of file, that error is io.EOF.